editor/components/font-sizes/with-font-sizes.js

```diff
@@ -0,0 +1,125 @@
+ *
+ * @return {Function} Higher-order component.
+ */
+export default ( ...fontSizeNames ) => {
+	// computes an object whose key is the font size attribute name as passed in the array,
+	// and the value is the custom font size attribute name.
+	// Custom font size is automatically compted by appending custom followed by the font size attribute name in with the first letter capitalized.
+	const fontSizeAttributeNames = reduce( fontSizeNames, ( fontSizeAttributeNamesAccumulator, fontSizeAttributeName ) => {
+		fontSizeAttributeNamesAccumulator[ fontSizeAttributeName ] = `custom${ upperFirst( fontSizeAttributeName ) }`;
+		return fontSizeAttributeNamesAccumulator;
+	}, {} );
+	return createHigherOrderComponent(
+		compose( [
+			withSelect( ( select ) => {
+				const { fontSizes } = select( 'core/editor' ).getEditorSettings();
+				return {
+					fontSizes,
+				};
+			} ),
+			( WrappedComponent ) => {
+				return class extends Component {
+					constructor( props ) {
+						super( props );
+
+						this.setters = this.createSetters();
+
+						this.state = {};
+					}
+
+					createSetters() {
+						return reduce( fontSizeAttributeNames, ( settersAccumulator, customFontSizeAttributeName, fontSizeAttributeName ) => {
+							const upperFirstFontSizeAttributeName = upperFirst( fontSizeAttributeName );
+							settersAccumulator[ `set${ upperFirstFontSizeAttributeName }` ] =
+								this.createSetFontSize( fontSizeAttributeName, customFontSizeAttributeName );
+							return settersAccumulator;
+						}, {} );
+					}
+
+					createSetFontSize( fontSizeAttributeName, customFontSizeAttributeName ) {
+						return ( fontSizeValue ) => {
+							const fontSizeObject = find( this.props.fontSizes, { size: fontSizeValue } );
+							this.props.setAttributes( {
+								[ fontSizeAttributeName ]: fontSizeObject && fontSizeObject.slug ? fontSizeObject.slug : undefined,
+								[ customFontSizeAttributeName ]: fontSizeObject && fontSizeObject.slug ? undefined : fontSizeValue,
+							} );
+						};
+					}
+
+					static getDerivedStateFromProps( { attributes, fontSizes }, previousState ) {
+						const didAttributesChange = ( customFontSizeAttributeName, fontSizeAttributeName ) => {
+							if ( previousState[ fontSizeAttributeName ] ) {
+								// if new font size is name compare with the previous slug
+								if ( attributes[ fontSizeAttributeName ] ) {
+									return attributes[ fontSizeAttributeName ] !== previousState[ fontSizeAttributeName ].slug;
+								}
+								// if font size is not named, update when the font size value changes.
+								return previousState[ fontSizeAttributeName ].size !== attributes[ customFontSizeAttributeName ];
+							}
+							// in this case we need to build the font size object
+							return true;
+						};
+
+						if ( ! some( fontSizeAttributeNames, didAttributesChange ) ) {
+							return previousState;
```

Comment:
It should return `null` when no changes happened. See: https://reactjs.org/blog/2018/03/27/update-on-async-rendering.html#new-lifecycle-getderivedstatefromprops.

Reply:
> The new static `getDerivedStateFromProps` lifecycle is invoked after a component is instantiated as well as before it is re-rendered. It can return an object to update state, or `null` to indicate that the new props do not require any state updates.

Reply:
Corrected, I'm still mind programmed to return the same value when changes did not happen.